File: frontend/src/services/api/terrain.ts
/**
 * Terrain/SRTM API Service
 *
 * Handles all terrain-related API calls:
 * - List terrain tiles
 * - Download SRTM tiles
 * - Query elevation
 * - Get coverage status
 * - Delete tiles
 */

import api from '@/lib/api';

// Type definitions
export interface TerrainTile {
  id?: string;
  tile_name: string;
  lat_min: number;
  lat_max: number;
  lon_min: number;
  lon_max: number;
  minio_bucket: string;
  minio_path: string;
  file_size_bytes?: number;
  status: 'pending' | 'downloading' | 'ready' | 'failed';
  error_message?: string;
  checksum_sha256?: string;
  source_url?: string;
  downloaded_at?: string;
  created_at?: string;
  updated_at?: string;
}

export interface TerrainTilesList {
  tiles: TerrainTile[];
  total: number;
  ready: number;
  downloading: number;
  failed: number;
  pending: number;
}

export interface BoundsRequest {
  lat_min: number;
  lat_max: number;
  lon_min: number;
  lon_max: number;
}

export interface DownloadRequest {
  bounds?: BoundsRequest;
}

export interface TileDownloadResult {
  tile_name: string;
  success: boolean;
  error?: string;
  file_size?: number;
  checksum?: string;
}

export interface DownloadResponse {
  successful: number;
  failed: number;
  total: number;
  tiles: TileDownloadResult[];
}

export interface CoverageRegion {
  lat_min: number;
  lat_max: number;
  lon_min: number;
  lon_max: number;
  tiles_needed: string[];
  tiles_ready: string[];
  tiles_downloading: string[];
  tiles_failed: string[];
  tiles_missing: string[];
}

export interface CoverageStatus {
  region: CoverageRegion;
  total_tiles: number;
  ready_count: number;
  downloading_count: number;
  failed_count: number;
  missing_count: number;
  coverage_percent: number;
}

export interface ElevationResponse {
  lat: number;
  lon: number;
  elevation_m: number;
  tile_name: string;
  source: 'srtm' | 'simplified';
}

/**
 * Terrain API client
 */
export const terrainApi = {
  /**
   * List all terrain tiles
   */
  async listTiles(): Promise<TerrainTilesList> {
    const response = await api.get('/v1/terrain/tiles');
    return response.data;
  },

  /**
   * Download terrain tiles
   * @param request Optional bounds, otherwise auto-detect from WebSDR stations
   */
  async downloadTiles(request?: DownloadRequest): Promise<DownloadResponse> {
    const response = await api.post('/v1/terrain/download', request || {});
    return response.data;
  },

  /**
   * Get coverage status for WebSDR region
   */
  async getCoverage(): Promise<CoverageStatus> {
    const response = await api.get('/v1/terrain/coverage');
    return response.data;
  },

  /**
   * Query elevation at specific coordinates
   */
  async queryElevation(lat: number, lon: number): Promise<ElevationResponse> {
    const response = await api.get('/v1/terrain/elevation', {
      params: { lat, lon }
    });
    return response.data;
  },

  /**
   * Delete a terrain tile
   */
  async deleteTile(tileName: string): Promise<{ success: boolean; message: string }> {
    const response = await api.delete(`/v1/terrain/tiles/${tileName}`);
    return response.data;
  }
};

File: services/backend/src/models/terrain.py
"""
Pydantic models for terrain/SRTM data management.
"""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class TerrainTile(BaseModel):
    """Single SRTM terrain tile information."""
    
    id: Optional[str] = None
    tile_name: str = Field(..., description="Tile name (e.g., 'N44E007')")
    lat_min: int = Field(..., description="Minimum latitude (SW corner)")
    lat_max: int = Field(..., description="Maximum latitude (NE corner)")
    lon_min: int = Field(..., description="Minimum longitude (SW corner)")
    lon_max: int = Field(..., description="Maximum longitude (NE corner)")
    
    minio_bucket: str = Field(default="heimdall-terrain")
    minio_path: str = Field(..., description="Path in MinIO bucket")
    file_size_bytes: Optional[int] = Field(None, description="File size in bytes")
    
    status: str = Field(..., description="Status: pending/downloading/ready/failed")
    error_message: Optional[str] = None
    
    checksum_sha256: Optional[str] = None
    source_url: Optional[str] = None
    
    downloaded_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class TerrainTilesList(BaseModel):
    """List of terrain tiles."""
    
    tiles: List[TerrainTile]
    total: int
    ready: int
    downloading: int
    failed: int
    pending: int


class BoundsRequest(BaseModel):
    """Geographic bounds for tile download."""
    
    lat_min: float = Field(..., ge=-90, le=90)
    lat_max: float = Field(..., ge=-90, le=90)
    lon_min: float = Field(..., ge=-180, le=180)
    lon_max: float = Field(..., ge=-180, le=180)


class DownloadRequest(BaseModel):
    """Request to download terrain tiles."""
    
    bounds: Optional[BoundsRequest] = Field(
        None,
        description="Optional geographic bounds. If not provided, auto-detect from WebSDR stations"
    )


class TileDownloadResult(BaseModel):
    """Result of a single tile download."""
    
    tile_name: str
    success: bool
    error: Optional[str] = None
    file_size: Optional[int] = None
    checksum: Optional[str] = None


class DownloadResponse(BaseModel):
    """Response from tile download operation."""
    
    successful: int
    failed: int
    total: int
    tiles: List[TileDownloadResult]


class CoverageRegion(BaseModel):
    """Coverage region information."""
    
    lat_min: float
    lat_max: float
    lon_min: float
    lon_max: float
    tiles_needed: List[str]
    tiles_ready: List[str]
    tiles_downloading: List[str]
    tiles_failed: List[str]
    tiles_missing: List[str]


class CoverageStatus(BaseModel):
    """Terrain coverage status."""
    
    region: CoverageRegion
    total_tiles: int
    ready_count: int
    downloading_count: int
    failed_count: int
    missing_count: int
    coverage_percent: float = Field(..., ge=0, le=100)


class ElevationQuery(BaseModel):
    """Query for elevation at specific coordinates."""
    
    lat: float = Field(..., ge=-90, le=90)
    lon: float = Field(..., ge=-180, le=180)


class ElevationResponse(BaseModel):
    """Elevation query response."""
    
    lat: float
    lon: float
    elevation_m: float
    tile_name: str
    source: str = Field(..., description="'srtm' or 'simplified'")
